### PR TITLE
ScrapeConfig CRD: Add LightSail Service Discovery Config Options

### LINKED ISSUE #6657: ScrapeConfig CRD: Add LightSail Service Discovery Configuration to The API
### Component(s)

_No response_

### What is missing? Please describe.

Add LightSail Service Discovery Configuration to The API

### Describe alternatives you've considered.

_No response_

## Environment
Kubernetes Version:
Prometheus-Operator Version:

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description
Add LightSail Service Discovery Config Options to the ScrapeConfig API.

Closes #6657

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```
Add LightSail Service Discovery Configuration Options to ScrapeConfig CRD
```

### CODE REVIEW

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 1]
```diff
@@ -1066,3 +1069,59 @@ type PuppetDBSDConfig struct {
 	// +optional
 	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
 }
+
+type LightSailSDConfig struct {
+	// The AWS region
+	// TODO: seems like an empty string is valid for the Region, should we make it Required
+	// and let it be, or leave it optional? considering that Prometheus does not
+	// check complain if it is nil.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1077]: if it's optional then you want a string pointer but ensure that the string length is >= 1 if not nil.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 2]
```diff
@@ -1066,3 +1069,59 @@ type PuppetDBSDConfig struct {
...
+	// +optional
+	Region *string `json:"region,omitempty"`
+	// AccessKey is the AWS API key.
+	// +optional
+	AccessKey *corev1.SecretKeySelector `json:"accessKey,omitempty"`
+	// SecretKey is the AWS API secret.
+	// +optional
+	SecretKey *corev1.SecretKeySelector `json:"secretKey,omitempty"`
+	// AWS Role ARN, an alternative to using AWS API keys.
+	// TODO: Should we validate against "^arn:aws:connect:\w+(?:-\w+)+:\d{12}:instance\/[A-Za-z0-9]+(?:-[A-Za-z0-9]+)+$"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1087]: Not sure about this one. Maybe tackle it later with AWS SD?

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 3]
```diff
@@ -1066,3 +1069,59 @@ type PuppetDBSDConfig struct {
...
+	// +optional
+	RoleARN *string `json:"roleARN,omitempty"`
+	// Custom endpoint to be used.
+	// TODO: Not obvious what validations are needed for this field.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1091]: the doc says "hostname only or fully qualified URI" so not a lot that we can do, expect rejecting empty string values.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 4]
```diff
@@ -1066,3 +1069,59 @@ type PuppetDBSDConfig struct {
...
+	// +optional
+	RefreshInterval *v1.Duration `json:"refreshInterval,omitempty"`
+	// Port to scrape the metrics from.
+	// If using the public IP address, this must instead be specified in the relabeling rule.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port *int32 `json:"port,omitempty"`
+	// Optional HTTP basic authentication information.
+	// Cannot be set at the same time as `authorization`, or `oauth2`.
+	// TODO: There is an extra "password_file" present in the SD config, should we support it?
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1105]: no, for security reasons (avoiding the exfiltration of credentials).

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 1 above)
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1077]: Added `MinLength=1`

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 2 above)
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1087]: I will leave the TODO, just in case. We can come back to it while refactoring.

Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1086]: The `TODO: ...` line appears in the API docs too which isn't very clean. I'll remove it for now.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -1066,3 +1069,54 @@ type PuppetDBSDConfig struct {
...
+	// SecretKey is the AWS API secret.
+	// +optional
+	SecretKey *corev1.SecretKeySelector `json:"secretKey,omitempty"`
+	// AWS Role ARN, an alternative to using AWS API keys.
+	// +optional
+	RoleARN *string `json:"roleARN,omitempty"`
+	// Custom endpoint to be used.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Endpoint string `json:"endpoint"`
```
[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1091]: Hmm.. but I also see that the endpoint field is not omitted either. Could it mean that we always expect it but do not necessarily need it to be non-empty?
I see the following:
1. Make it optional.
2. Keep the `MinLength` validation.
3. Do not add the omitempty tag.

Would that be good?

[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1091]: You want optional + min length >= 1 + omitempty, all meaning the field is optional but if defined, it shouldn't be the empty string.

```suggestion
	// +kubebuilder:validation:MinLength=1
	// +optional
	Endpoint *string `json:"endpoint,omitempty"`
```

[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1091]: Got it!

**Review by simonpasquier — APPROVED** [review]
lgtm

**Pull request merged**